File: src/votes/votes.controller.ts
import {
  Controller, Get, Post, Put, Delete,
  Param, Body, Query, UseGuards, Request,
} from '@nestjs/common';
import { JwtAuthGuard } from '../auth/guards/jwt.guard';
import { VotesService } from './votes.service';

@Controller('vote-sessions')
@UseGuards(JwtAuthGuard)
export class VotesController {
  constructor(private readonly votesService: VotesService) {}

  @Get()
  getSession(@Request() req: any, @Query('playlistId') playlistId: string) {
    return this.votesService.getSession(req.user.bandId, playlistId);
  }

  @Post()
  createSession(
    @Request() req: any,
    @Body() body: { playlistId: string; title: string },
  ) {
    return this.votesService.createSession(req.user.bandId, body.playlistId, body.title);
  }

  @Put(':id/close')
  closeSession(@Request() req: any, @Param('id') id: string) {
    return this.votesService.closeSession(req.user.bandId, id);
  }

  @Put(':id/reopen')
  reopenSession(@Request() req: any, @Param('id') id: string) {
    return this.votesService.reopenSession(req.user.bandId, id);
  }

  @Delete(':id')
  deleteSession(@Request() req: any, @Param('id') id: string) {
    return this.votesService.deleteSession(req.user.bandId, id);
  }

  @Post(':id/votes')
  castVote(
    @Request() req: any,
    @Param('id') id: string,
    @Body() body: { voterName: string; orderedIds: string[] },
  ) {
    return this.votesService.castVote(req.user.bandId, id, body.voterName, body.orderedIds);
  }

  @Get(':id/results')
  getResults(@Request() req: any, @Param('id') id: string) {
    return this.votesService.getResults(req.user.bandId, id);
  }
}

File: spring-src/main/java/com/blackout/api/votes/infrastructure/web/VotesController.java
package com.blackout.api.votes.infrastructure.web;

import com.blackout.api.shared.infrastructure.security.BlackoutAuthentication;
import com.blackout.api.votes.application.service.VotesApplicationService;
import com.blackout.api.votes.infrastructure.web.dto.*;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/vote-sessions")
public class VotesController {

    private final VotesApplicationService service;

    public VotesController(VotesApplicationService service) {
        this.service = service;
    }

    @GetMapping
    public ResponseEntity<VoteSessionResponse> getSession(
            BlackoutAuthentication auth,
            @RequestParam String playlistId
    ) {
        return service.getSession(auth.getBandId(), playlistId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.ok(null));
    }

    @PostMapping
    public VoteSessionResponse createSession(
            BlackoutAuthentication auth,
            @RequestBody CreateSessionRequest request
    ) {
        return service.createSession(auth.getBandId(), request.playlistId(), request.title());
    }

    @RequestMapping(value = "/{id}/close", method = {RequestMethod.PUT, RequestMethod.POST})
    public VoteSessionResponse closeSession(
            BlackoutAuthentication auth,
            @PathVariable String id
    ) {
        return service.closeSession(auth.getBandId(), id);
    }

    @RequestMapping(value = "/{id}/reopen", method = {RequestMethod.PUT, RequestMethod.POST})
    public VoteSessionResponse reopenSession(
            BlackoutAuthentication auth,
            @PathVariable String id
    ) {
        return service.reopenSession(auth.getBandId(), id);
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteSession(
            BlackoutAuthentication auth,
            @PathVariable String id
    ) {
        service.deleteSession(auth.getBandId(), id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/{id}/votes")
    public VoteResponse castVote(
            BlackoutAuthentication auth,
            @PathVariable String id,
            @RequestBody CastVoteRequest request
    ) {
        return service.castVote(auth.getBandId(), id, request.voterName(), request.orderedIds());
    }

    @GetMapping("/{id}/results")
    public List<VoteResultEntry> getResults(
            BlackoutAuthentication auth,
            @PathVariable String id
    ) {
        return service.getResults(auth.getBandId(), id);
    }
}
